File: laravel/compare-html.py
#!/usr/bin/env python3
"""
HTML Body Structure Comparator
Compares body HTML (excluding scripts) between two URLs.
Usage: python3 compare-html.py [url1] [url2]
"""
import sys
import urllib.request
from html.parser import HTMLParser

URL_REF = "https://italia.github.io/design-comuni-pagine-statiche/sito/homepage.html"
URL_LOCAL = "http://127.0.0.1:8000/it/tests/homepage"


class BodyExtractor(HTMLParser):
    def __init__(self):
        super().__init__()
        self.in_body = False
        self.in_script = False
        self.depth = 0
        self.elements = []  # list of (tag, attrs_dict)

    def handle_starttag(self, tag, attrs):
        if tag == "body":
            self.in_body = True
            return
        if not self.in_body:
            return
        if tag == "script":
            self.in_script = True
            return
        if not self.in_script:
            attrs_dict = {k: v for k, v in attrs if k not in ("class", "style", "id")}
            self.elements.append(("open", tag, attrs_dict))

    def handle_endtag(self, tag):
        if tag == "script":
            self.in_script = False
            return
        if not self.in_body or self.in_script:
            return
        if tag == "body":
            self.in_body = False
            return
        self.elements.append(("close", tag))

    def handle_data(self, data):
        if not self.in_body or self.in_script:
            return
        text = data.strip()
        if text:
            self.elements.append(("text", text[:80]))


def fetch(url):
    req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(req, timeout=15) as r:
        return r.read().decode("utf-8", errors="replace")


def extract(html):
    p = BodyExtractor()
    p.feed(html)
    return p.elements


def tag_sequence(elements):
    """Return just open tags in order (structural skeleton)."""
    return [f"<{tag}>" for kind, tag, *_ in elements if kind == "open"]


def compare(url1, url2):
    print(f"Fetching REF:   {url1}")
    html1 = fetch(url1)
    print(f"Fetching LOCAL: {url2}")
    html2 = fetch(url2)

    e1 = extract(html1)
    e2 = extract(html2)

    s1 = tag_sequence(e1)
    s2 = tag_sequence(e2)

    print(f"\nREF   tags: {len(s1)}")
    print(f"LOCAL tags: {len(s2)}")

    # Find differences
    set1 = {}
    set2 = {}
    for t in s1:
        set1[t] = set1.get(t, 0) + 1
    for t in s2:
        set2[t] = set2.get(t, 0) + 1

    all_tags = sorted(set(set1) | set(set2))
    diffs = [(t, set1.get(t, 0), set2.get(t, 0)) for t in all_tags if set1.get(t, 0) != set2.get(t, 0)]

    if not diffs:
        print("\n✅ TAG COUNTS MATCH - same structural skeleton")
    else:
        print(f"\n❌ {len(diffs)} TAG COUNT DIFFERENCES:")
        print(f"{'TAG':<30} {'REF':>6} {'LOCAL':>6} {'DIFF':>6}")
        print("-" * 50)
        for tag, c1, c2 in sorted(diffs, key=lambda x: abs(x[1]-x[2]), reverse=True):
            diff = c2 - c1
            sign = "+" if diff > 0 else ""
            print(f"{tag:<30} {c1:>6} {c2:>6} {sign}{diff:>5}")

    # Sequential diff (first N mismatches)
    print("\n--- SEQUENTIAL STRUCTURE DIFF (first 20 mismatches) ---")
    mismatches = 0
    for i, (t1, t2) in enumerate(zip(s1, s2)):
        if t1 != t2:
            print(f"  pos {i:4d}: REF={t1:<25} LOCAL={t2}")
            mismatches += 1
            if mismatches >= 20:
                print("  ... (truncated)")
                break
    if mismatches == 0:
        print("  ✅ Sequential order matches (up to min length)")

    # Length diff
    if len(s1) != len(s2):
        extra = len(s1) - len(s2)
        if extra > 0:
            print(f"\n  REF has {extra} extra tags at end: {s1[len(s2):len(s2)+10]}")
        else:
            print(f"\n  LOCAL has {-extra} extra tags at end: {s2[len(s1):len(s1)+10]}")

    return len(diffs) == 0 and len(s1) == len(s2)


if __name__ == "__main__":
    url1 = sys.argv[1] if len(sys.argv) > 1 else URL_REF
    url2 = sys.argv[2] if len(sys.argv) > 2 else URL_LOCAL
    ok = compare(url1, url2)
    sys.exit(0 if ok else 1)

File: laravel/compare-html.php
#!/usr/bin/env php
<?php
/**
 * HTML Body Comparison Tool
 * Compares HTML structure between two URLs (excluding scripts)
 * 
 * Usage: php compare-html.php <url1> <url2>
 */

$url1 = $argv[1] ?? 'https://italia.github.io/design-comuni-pagine-statiche/sito/homepage.html';
$url2 = $argv[2] ?? 'http://127.0.0.1:8000/it/tests/homepage';

// Fetch HTML using curl for better compatibility
function fetchHtml($url) {
    $ch = curl_init($url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_TIMEOUT, 10);
    curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
    $html = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);
    
    if ($httpCode !== 200 || !$html) {
        return null;
    }
    return $html;
}

echo "🔍 HTML Body Comparison Tool\n";
echo "Originale: $url1\n";
echo "Replica: $url2\n\n";
echo "⏳ Download pagine...\n";

// Fetch HTML
$html1 = fetchHtml($url1);
$html2 = fetchHtml($url2);

if (!$html1) {
    echo "❌ Errore: Impossibile scaricare originale (HTTP " . (curl_init($url1) ? 'timeout' : 'error') . ")\n";
    exit(1);
}

if (!$html2) {
    echo "❌ Errore: Impossibile scaricare replica (HTTP timeout o errore)\n";
    exit(1);
}

echo "✅ Pagine scaricate\n";
echo "Originale: " . strlen($html1) . " bytes\n";
echo "Replica: " . strlen($html2) . " bytes\n\n";

// Extract body content
preg_match('/<body[^>]*>(.*?)<\/body>/s', $html1, $matches1);
preg_match('/<body[^>]*>(.*?)<\/body>/s', $html2, $matches2);

$body1 = $matches1[1] ?? '';
$body2 = $matches2[1] ?? '';

// Remove scripts
$body1 = preg_replace('/<script\b[^>]*>.*?<\/script>/is', '', $body1);
$body2 = preg_replace('/<script\b[^>]*>.*?<\/script>/is', '', $body2);

// Extract structure (tags with classes/IDs)
function extractStructure($html) {
    preg_match_all('/<(\/?)([a-z0-9]+)([^>]*)(\/?)>/i', $html, $matches, PREG_SET_ORDER);
    $structure = [];
    foreach ($matches as $match) {
        $tag = $match[2];
        $attrs = $match[3];
        
        // Extract class
        preg_match('/class=["\']([^"\']*)["\']/i', $attrs, $classMatch);
        $class = $classMatch[1] ?? '';
        
        // Extract id
        preg_match('/id=["\']([^"\']*)["\']/i', $attrs, $idMatch);
        $id = $idMatch[1] ?? '';
        
        $structure[] = [
            'close' => $match[1] === '/',
            'tag' => $tag,
            'class' => $class,
            'id' => $id,
        ];
    }
    return $structure;
}

$struct1 = extractStructure($body1);
$struct2 = extractStructure($body2);

// Compare
echo "📊 Confronto Struttura:\n";
echo "Originale: " . count($struct1) . " elementi\n";
echo "Replica: " . count($struct2) . " elementi\n\n";

$differences = [];
$maxLen = max(count($struct1), count($struct2));

for ($i = 0; $i < $maxLen; $i++) {
    $elem1 = $struct1[$i] ?? null;
    $elem2 = $struct2[$i] ?? null;
    
    if (!$elem1 && $elem2) {
        $differences[] = "➕ EXTRA in replica: <{$elem2['tag']} class=\"{$elem2['class']}\" id=\"{$elem2['id']}\">";
    } elseif ($elem1 && !$elem2) {
        $differences[] = "➖ MANCA in replica: <{$elem1['tag']} class=\"{$elem1['class']}\" id=\"{$elem1['id']}\">";
    } elseif ($elem1 && $elem2) {
        if ($elem1['tag'] !== $elem2['tag']) {
            $differences[] = "🏷️ TAG DIVERSO: {$elem1['tag']} vs {$elem2['tag']}";
        }
        if ($elem1['class'] !== $elem2['class']) {
            $differences[] = "🎨 CLASSE DIVERSA: {$elem1['class']} vs {$elem2['class']}";
        }
        if ($elem1['id'] !== $elem2['id']) {
            $differences[] = "🆔 ID DIVERSO: {$elem1['id']} vs {$elem2['id']}";
        }
    }
}

if (empty($differences)) {
    echo "✅ STRUTTURA IDENTICA!\n";
} else {
    echo "❌ Differenze trovate:\n\n";
    foreach (array_slice($differences, 0, 50) as $diff) {
        echo "$diff\n";
    }
    if (count($differences) > 50) {
        echo "\n... e altre " . (count($differences) - 50) . " differenze\n";
    }
}

echo "\n📈 Statistiche:\n";
echo "Differenze totali: " . count($differences) . "\n";
$matchPercent = 100 - (count($differences) / max(count($struct1), count($struct2)) * 100);
echo "Corrispondenza: " . round($matchPercent, 2) . "%\n";
